---
title: "Infer ASV's With DADA2"
output: html_document
date: "2024-02-21"
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, 
                      fig.align = "center", 
                      fig.path = "../figures/01_DADA2/") # send any figure output to this folder
```

# Before you start
```{r set-seed}
set.seed(238428)
```


# Goals of this file 

1. Use raw fastq files and generate quality plots to assess quality of reads
2. Filter and trim out bad sequences and bases from our sequencing files.
3. Write out fastq files with high quality sequences.
4. Evaluate the quality from our filter and trim.

# Load libraries

```{r load libraries}
install.packages("BiocManager")
library(devtools)

BiocManager::install("dada2")
library(dada2)

library(ggplot2)

install.packages("tidyverse")
library(tidyverse)
```

# Load Data
```{r}
# Set the row fastq path to the raw sequencing files
# Path to the fastq files
raw_fastqs_path <- "data/01_DADA2/01_rawgzipped_fastqs"
raw_fastqs_path

# What files are in this path? Intuition Check
list.files(raw_fastqs_path)

# How many files are there?
str(list.files(raw_fastqs_path))


# Create a vector of forward reads
forward_reads <- list.files(raw_fastqs_path, pattern = "R1_001.fastq.gz", full.names = TRUE)

#Final Intuition Check
head(forward_reads)

# Create a vector of reverse reads

reverse_reads <- list.files(raw_fastqs_path, pattern = "R2_001.fastq.gz", full.names = TRUE)

#Final Intuition Check
head(reverse_reads)
```

# Quality Plots
```{r raw-quality-plot}
#Randomly select 2 samples from dataset to evaluate
random_samples <-sample(1:length(reverse_reads), size = 2)
random_samples

# Calculate and plot quality of these two samples
plotQualityProfile(forward_reads[random_samples])
plotQualityProfile(reverse_reads[random_samples])
```


# Placeholder for filtered reads

```{r prep-filered-reads}

# Make vector of our samples, extract sample name from files

samples <- 
 sapply(strsplit(basename(forward_reads), "_"), `[`,1) 

#strsplit splits everything into a vector

#Place filtered reads into filtered_fastq_path
filtered_fastqs_path <- "data/01_DADA2/02_filtered_fastqs"

#create 2 variables: filtered_F, filtered_R
filtered_forward_reads <- file.path(filtered_fastqs_path, paste0(samples, "_R1_filtered.fastq.gz"))
length(filtered_forward_reads)


#create vector with sample names

# reverse reads
filtered_reverse_reads <- file.path(filtered_fastqs_path, paste0(samples, "_R2_filtered.fastq.gz"))
length(filtered_reverse_reads)
```

# Filter and Trim Reads

```{r filter-and-trim}
?filterAndTrim
filtered_reads <- 
filterAndTrim(fwd = forward_reads, filt = filtered_forward_reads,
       rev = reverse_reads, filt.rev = filtered_reverse_reads,
       maxN = 0, maxEE = c(2,2), trimLeft = 3, 
       truncQ = 2, rm.phix = TRUE, compress = TRUE,
       # multithread = TRUE # DO NOT USE THIS ARGUMENT IN CLASS
)
```


```{r filterTrim-quality-plots}
plotQualityProfile(filtered_forward_reads[random_samples]) + 
labs(title = "Trimmed Forward Read Quality")

plotQualityProfile(filtered_reverse_reads[random_samples])+ 
labs(title = "Trimmed Reverse Read Quality")
```

## Stats on Read Output from `filterAndTrim`
```{r filterTrim-stats}
dplyr::reframe
filtered_df <- as.data.frame(filtered_reads)
head(filtered_df)

# calculate some stats
filtered_df %>%
reframe(median_reads_in = median(reads.in),
               median_reads_out = median(reads.out),
               median_percent_retained = median(reads.out)/median(reads.in))
```

# Error Modeling
```{r learn-errors}
# Forward reads 
error_forward_reads <- 
  learnErrors(filtered_forward_reads) # multithread = TRUE
# Plot Forward  
plotErrors(error_forward_reads, nominalQ = TRUE) + 
  labs(title = "Forward Read Error Model")

# Reverse reads 
error_reverse_reads <- 
  learnErrors(filtered_reverse_reads) # multithread = TRUE
# Plot reverse
plotErrors(error_reverse_reads, nominalQ = TRUE) + 
  labs(title = "Reverse Read Error Model")
```

Infer ASVs
```{r infer-ASVs}
# Infer forward ASVs
dada_forward <- dada(filtered_forward_reads,
                     err = error_forward_reads,  multithread = 4)

# Infer reverse ASVs
dada_reverse <- dada(filtered_reverse_reads,
                     err = error_reverse_reads,  multithread = 4)
```

# Merge Forward & Reverse ASVs
```{r merge-ASVs}
# merge forward and reverse ASVs
merged_ASVs <- mergePairs(dada_forward, filtered_forward_reads, 
                          dada_reverse, filtered_reverse_reads,
                          verbose = TRUE)

# Evaluate the output 
typeof(merged_ASVs)
length(merged_ASVs)
names(merged_ASVs)
```

```{r generate-ASV-table}
# Create the ASV Count Table 
raw_ASV_table <- makeSequenceTable(merged_ASVs)

# Write out the file to data/01_DADA2
```


# Session Information

```{r session-info}
# Ensure reproducibility
devtools::session_info()
```

